### ISSUE TITLE
Minimizing M2 calls in parser

### ISSUE THREAD

**Issue by sommars** [musicman3320/m2r#79]
Suppose we wanted to retrieve two rings R and S from M2, as could happen during m2_parse(). What if we called M2 once with the input of (R,S)? Would this improve speed?

**Comment by musicman3320** [comment]
This is unlikely to be easy to implement, as it requires the parser to peek ahead and grab several rings at once.  After a recent update to `m2()`, subsequent calls are much faster.

**Comment by sommars** [comment]
No, peaking ahead isn't necessary. Instead of getting them on the fly, wait til you finish parsing, then grab all of them at once.

**Comment by musicman3320** [comment]
But the value returned by `m2()` determines how they are parsed, and thus how subsequent parsing occurs.  For instance, `R^2` becomes a module if `R` is a ring, and an expression if `R` is a variable.  

It is possible this can be done in a lazier fashion than it is right now, but I don't see a way to do so without significantly modifying the parser (both physically and philosophically).  Perhaps we can reopen if it becomes a bottleneck in the future?

**Comment by sommars** [comment]
Ohhhh, right. Yeah, that's true

**Comment by musicman3320** [comment]
It is certainly worth looking into ways of minimizing the calls to `m2()` when parsing.  Some ideas: 

1. Keep a list of known function/class names that will not need to be looked up.  
2. Check the format of the name (e.g. `m2rintring0000000000001`) and only look up those that don't fit.   

Both of these can be subverted by a malitious user, but at some point we need to trust them.  I will open a new bug so that we can (potentially) add this in a future version.

**Comment by coneill-math** [comment]
Excess calls to `m2()` in the parser are more evident in the cloud version.  

We should bump the priority on this.
